src/util/mod.rs

```diff
@@ -0,0 +1,100 @@
+use rand::prelude::*;
+use std::collections::HashMap;
+use std::collections::LinkedList;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+const BACKOFF_TIME_MAX_MS: f32 = 24f32 * 60f32 * 60f32 * 1000f32;
+const BACKOFF_POW: f32 = 1.05f32;
+const BACKOFF_TIME_MS: f32 = 5f32 * 1000f32;
+
+#[napi]
+
+pub struct QueueWithBackoff {
+    queue: LinkedList<String>,
+    backoff: HashMap<String, u128>,
+    last_backoff: HashMap<String, u32>,
+}
+
+impl Default for QueueWithBackoff {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+#[napi]
+
+impl QueueWithBackoff {
+    #[napi(constructor)]
+    pub fn new() -> Self {
+        QueueWithBackoff {
+            queue: LinkedList::new(),
+            backoff: HashMap::new(),
+            last_backoff: HashMap::new(),
+        }
+    }
+
+    #[napi]
+    pub fn pop(&mut self) -> Option<String> {
+        let start = SystemTime::now();
+        let since_the_epoch = start.duration_since(UNIX_EPOCH).unwrap();
+
+        let mut items_to_rm: Vec<String> = vec![];
+        for item in self.backoff.iter() {
+            if *item.1 < since_the_epoch.as_millis() {
+                self.queue.push_back(item.0.clone());
+                items_to_rm.push(item.0.clone());
+            }
+        }
+
+        for item in items_to_rm {
+            self.backoff.remove(&item);
+        }
```

Comment:
Something about this strikes me as inefficient. I sort of want to effectively pop only some items from the HashMap. Currently we have to do a O(N) iteration on each pop, which is pretty rubbish.